### ISSUE TITLE
OOM error deploying package

### ISSUE THREAD

**Issue by jcdesrochers** [sapia-oss/corus#30]
Corus version: 5.0.3

Deploying a package of 145 MB gives the folowing stack trace:

``2016.06.03@15:38:32:753 ERROR[org.sapia.corus.taskmanager.CorusTaskManagerImpl]: DeployTask >> Problem occurred executing task java.lang.OutOfMemoryError: Java heap space
        at org.sapia.corus.client.common.ZipUtil.readBytes(ZipUtil.java:128)
        at org.sapia.corus.client.common.ZipUtil.readEntry(ZipUtil.java:43)
        at org.sapia.corus.client.common.ZipUtil.readEntryStream(ZipUtil.java:106)
        at org.sapia.corus.client.common.ZipUtil.readEntryStream(ZipUtil.java:75)
        at org.sapia.corus.file.FileSystemModuleImpl.openZipEntryStream(FileSystemModuleImpl.java:97)
        at org.sapia.corus.client.services.deployer.dist.Distribution.newInstance(Distribution.java:358)
        at org.sapia.corus.deployer.task.DeployTask.doExecute(DeployTask.java:109)
        at org.sapia.corus.deployer.task.DeployTask.execute(DeployTask.java:82)
        at org.sapia.corus.deployer.task.DeployTask.execute(DeployTask.java:73)
        at org.sapia.corus.taskmanager.core.TaskManagerImpl$1.run(TaskManagerImpl.java:53)
        at org.sapia.corus.taskmanager.core.SemaphoreThrottle.execute(SemaphoreThrottle.java:36)
        at org.sapia.corus.taskmanager.core.TaskManagerImpl$4.run(TaskManagerImpl.java:181)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)``

Further along the logs, found some other OOM error... is there's a memory leak?

``2016.06.03@16:05:09:108 ERROR[org.sapia.corus.taskmanager.CorusTaskManagerImpl]: ProcessCheckTask >> Problem occurred executing task java.lang.OutOfMemoryError: Java heap space
2016.06.03@16:05:18:701 ERROR[org.sapia.corus.taskmanager.CorusTaskManagerImpl]: ProcessCheckTask >> Problem occurred executing task java.lang.OutOfMemoryError: Java heap space
2016.06.03@16:05:35:942 ERROR[org.sapia.corus.taskmanager.CorusTaskManagerImpl]: ProcessCheckTask >> Problem occurred executing task java.lang.OutOfMemoryError: Java heap space
``

**Comment by jcdesrochers** [comment]
Clarification on the bug:

1. This happened on two servers at the same time (both part of the same cluster)
2. After a corus restart (of both server), the deployment worked like a charm

**Comment by jcdesrochers** [comment]
Another case of OutOfMemory error that happened 6 days after the initial startup:

```
[2016-06-08 00:34:08:989] SLF4J: Class path contains multiple SLF4J bindings.
[2016-06-08 00:34:08:989] SLF4J: Found binding in [jar:file:/opt/corus/5.0.4/lib/server/slf4j-jdk14-1.7.10.jar!/org/slf4j/impl/StaticLoggerBinder.class]
[2016-06-08 00:34:08:989] SLF4J: Found binding in [jar:file:/opt/corus/5.0.4/lib/server/avis-client-1.2.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
[2016-06-08 00:34:08:989] SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
[2016-06-08 00:34:08:991] SLF4J: Actual binding is of type [org.slf4j.impl.JDK14LoggerFactory]
[2016-06-08 00:34:10:465] Corus server (5.0.4) started on: http://10.2.71.152:33000/ubik, domain: geo-rt-int01
[2016-06-14 04:37:05:483] [2016/06/14 04:37:05:483][org.sapia.ubik.mcast.tcp.mina.MinaTcpUnicastHandler@SocketAcceptorIoProcessor-0.2] [ERROR]Exception caught - Already released buffer.  You released the buffer too many times.
[2016-06-14 04:37:05:483] java.lang.IllegalStateException: Already released buffer.  You released the buffer too many times.
[2016-06-14 04:37:05:484]       at org.apache.mina.common.PooledByteBufferAllocator$PooledByteBuffer.release(PooledByteBufferAllocator.java:310)
[2016-06-14 04:37:05:484]       at org.apache.mina.common.ByteBufferProxy.release(ByteBufferProxy.java:69)
[2016-06-14 04:37:05:484]       at org.apache.mina.transport.socket.nio.SocketIoProcessor.releaseWriteBuffers(SocketIoProcessor.java:350)
[2016-06-14 04:37:05:484]       at org.apache.mina.transport.socket.nio.SocketIoProcessor.doFlush(SocketIoProcessor.java:314)
[2016-06-14 04:37:05:484]       at org.apache.mina.transport.socket.nio.SocketIoProcessor.access$500(SocketIoProcessor.java:45)
[2016-06-14 04:37:05:484]       at org.apache.mina.transport.socket.nio.SocketIoProcessor$Worker.run(SocketIoProcessor.java:488)
[2016-06-14 04:37:05:484]       at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:51)
[2016-06-14 04:37:05:484]       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[2016-06-14 04:37:05:484]       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[2016-06-14 04:37:05:484]       at java.lang.Thread.run(Thread.java:745)
Exception in thread "Notifier: Thread-19" [2016-06-15 01:20:09:864] java.lang.OutOfMemoryError: Java heap space

Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "Notifier: Thread-19"
Exception in thread "Notifier: Thread-25"
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "Notifier: Thread-25"
Exception in thread "ActionDistributor: Thread-11" [2016-06-15 01:20:23:851] java.lang.OutOfMemoryError: Java heap space
Exception in thread "Ubik.EventChannel.Timer" [2016-06-15 01:20:40:963] java.lang.OutOfMemoryError: Java heap space
Exception in thread "ActionDistributor: Thread-14" Exception in thread "ActionDistributor: Thread-9" [2016-06-15 01:20:44:130] java.lang.OutOfMemoryError: Java heap space
[2016-06-15 01:20:45:372] java.lang.OutOfMemoryError: Java heap space
[2016-06-15 01:27:35:705] [2016/06/15 01:27:35:705][org.sapia.ubik.mcast.tcp.mina.MinaTcpUnicastHandler@AnonymousIoService-5] [ERROR]Exception caught - Java heap space
[2016-06-15 01:27:35:705] java.lang.OutOfMemoryError: Java heap space
[2016-06-15 01:27:35:910] [2016/06/15 01:27:35:910][org.sapia.ubik.mcast.tcp.mina.MinaTcpUnicastHandler@AnonymousIoService-11] [ERROR]Exception caught - Java heap space
[2016-06-15 01:27:35:910] java.lang.OutOfMemoryError: Java heap space
[2016-06-15 01:30:50:419] [2016/06/15 01:30:50:419][org.sapia.ubik.mcast.tcp.mina.MinaTcpUnicastHandler@AnonymousIoService-3] [ERROR]Exception caught - Java heap space
[2016-06-15 01:30:50:419] java.lang.OutOfMemoryError: Java heap space
[2016-06-15 01:30:54:425] [2016/06/15 01:30:54:425][org.sapia.ubik.mcast.tcp.mina.MinaTcpUnicastHandler@AnonymousIoService-12] [ERROR]Exception caught - Java heap space
[2016-06-15 01:30:54:425] java.lang.OutOfMemoryError: Java heap space
[2016-06-15 01:30:54:425] [2016/06/15 01:30:54:425][org.sapia.ubik.mcast.tcp.mina.MinaTcpUnicastHandler@AnonymousIoService-12] [ERROR]Exception caught - java.lang.IllegalStateException: doDecode() can't return true when buffer is not consumed. (Hexdump: 00 00 04 72)
[2016-06-15 01:30:54:425] org.apache.mina.filter.codec.ProtocolDecoderException: java.lang.IllegalStateException: doDecode() can't return true when buffer is not consumed. (Hexdump: 00 00 04 72)
[2016-06-15 01:30:54:425]       at org.apache.mina.filter.codec.ProtocolCodecFilter.messageReceived(ProtocolCodecFilter.java:165)
[2016-06-15 01:30:54:425]       at org.apache.mina.common.support.AbstractIoFilterChain.callNextMessageReceived(AbstractIoFilterChain.java:299)
[2016-06-15 01:30:54:425]       at org.apache.mina.common.support.AbstractIoFilterChain.access$1100(AbstractIoFilterChain.java:53)
[2016-06-15 01:30:54:425]       at org.apache.mina.common.support.AbstractIoFilterChain$EntryImpl$1.messageReceived(AbstractIoFilterChain.java:648)
[2016-06-15 01:30:54:425]       at org.apache.mina.filter.executor.ExecutorFilter.processEvent(ExecutorFilter.java:220)
[2016-06-15 01:30:54:425]       at org.apache.mina.filter.executor.ExecutorFilter$ProcessEventsRunnable.run(ExecutorFilter.java:264)
[2016-06-15 01:30:54:425]       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[2016-06-15 01:30:54:425]       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[2016-06-15 01:30:54:425]       at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:51)
[2016-06-15 01:30:54:425]       at java.lang.Thread.run(Thread.java:745)
[2016-06-15 01:30:54:426] Caused by: java.lang.IllegalStateException: doDecode() can't return true when buffer is not consumed.
[2016-06-15 01:30:54:426]       at org.apache.mina.filter.codec.CumulativeProtocolDecoder.decode(CumulativeProtocolDecoder.java:136)
```

**Comment by jcdesrochers** [comment]
JMX status of a corus node that has been running for one week. The memory used clearly indicates a memory leak somewhere!!!

```
<vmStatus domain="geo-rt-int01" host="10.2.71.165" port="33000" creationTime="1466003060177" creationDate="Wed Jun 15 15:04:20 UTC 2016">
<attribute name="runtime.startTime" value="1465314825717"/>
<attribute name="runtime.startDate" value="Tue Jun 07 15:53:45 UTC 2016"/>
<attribute name="runtime.upTime" value="688233974"/>
<attribute name="mem.pendingFinalization.count" value="0"/>
<attribute name="mem.heap.init" value="33554432"/>
<attribute name="mem.heap.committed" value="97320960"/>
<attribute name="mem.heap.max" value="97320960"/>
<attribute name="mem.heap.used" value="89519712"/>
<attribute name="mem.non-heap.init" value="2555904"/>
<attribute name="mem.non-heap.committed" value="62521344"/>
<attribute name="mem.non-heap.max" value="-1"/>
<attribute name="mem.non-heap.used" value="60800408"/>
<attribute name="gc.count" value="49313"/>
<attribute name="gc.elapsedTime" value="1379437"/>
<attribute name="threads.count" value="108"/>
<attribute name="threads.daemon.count" value="54"/>
<attribute name="threads.peak.count" value="110"/>
<attribute name="threads.totalStarted.count" value="57215"/>
<attribute name="classes.loadedCount" value="5416"/>
<attribute name="class.totalLoadedCount" value="7997"/>
<attribute name="classes.unloadedCount" value="2581"/>
<attribute name="os.name" value="Linux"/>
<attribute name="os.version" value="4.4.10-22.54.amzn1.x86_64"/>
<attribute name="os.arch" value="amd64"/>
<attribute name="os.processors" value="4"/>
</vmStatus>
```

**Comment by jcdesrochers** [comment]
Fixed with commit 5c57411268335b022ee382c88ba4a80661e97aa5
